File: python/vb2pyconv/vb2py/PythonCard/samples/worldclock/xearth.js

// KEA
// The two main functions use global variables, so we have to
// set all this up unless the functions are going to get changed.

var ClientRecieveTime=new Date();

var QueryTimeZone = -7;
var QueryTimeZoneOffsetMin = QueryTimeZone * 60 // -420

var NISTSendTimeGMTms = ClientRecieveTime.getTime() - ( QueryTimeZoneOffsetMin * 60 * 1000 );


// NIST start time in the query time zone
var QueryTimeZoneOffset = ( QueryTimeZoneOffsetMin * 60 * 1000 ) ;

// Client start time in some time zone
var ClientRecieveTimems = ClientRecieveTime.getTime() ;

// what timezone does your computer think it is? - in minutes
//var ClientTimeZone = Math.floor(NISTSendTime.getTimezoneOffset()) ;
var ClientTimeZone = QueryTimeZoneOffsetMin;

var ClientNISTDelta = Math.floor(NISTSendTimeGMTms - ClientRecieveTimems);

//WScript.echo(QueryTimeZoneOffset)
//WScript.echo(" ")
//WScript.echo(ClientRecieveTime)
//WScript.echo(" ")
//WScript.echo(ClientTimeZone)
//WScript.echo(" ")
//WScript.echo(ClientNISTDelta)
//WScript.echo(" ")

WScript.echo(updatexearthImage());


function updatexearthImage() {

  // get the computer's date
  var Thisdate = new Date();
  // convert to NIST date, using offset calculated in ClientNISTDelta
  ThisMilliseconds =  Thisdate.getTime() + ClientNISTDelta ;

  // KEA get rid of reference to DOM, just return the data
  // document.images['xearth'].src = xearthSrc(ThisMilliseconds);
  return xearthSrc(ThisMilliseconds);

  // setTimeout("updatexearthImage()", 10000);
}


function xearthSrc(ThisMilliseconds) {

var ssue = ThisMilliseconds / 1000;
var TWOPI = 2 * Math.PI;
var EpochStart = 631065600;
var DaysSinceEpoch = (ssue - EpochStart) / (24*3600);

var RadsPerDay = TWOPI / 365.242191;

var Epsilon_g = 279.403303 * (TWOPI / 360);
var OmegaBar_g = 282.768422 * (TWOPI / 360);
var Eccentricity = 0.016713;

var MeanObliquity = 23.440592 * (TWOPI / 360);

//Compute lambda = sun_ecliptic_longitude(ssue)
//Begin sun_ecliptic_longitude
  var D = DaysSinceEpoch;
  var N = RadsPerDay * D;
  N = N % TWOPI;
  if (N < 0) N += TWOPI;

  var M_sun = N + Epsilon_g - OmegaBar_g;
  if (M_sun < 0) M_sun += TWOPI;

  //Compute var E = solve_keplers_equation(M_sun);
  //Begin solve_keplers_equation
    var E = M_sun;
    var delta;
    while (1) {
      delta = E - (Eccentricity*Math.sin(E)) - M_sun;
      if (Math.abs(delta) <= 1E-10) break;
      E -= delta / (1 - (Eccentricity*Math.cos(E)));
    }
  //End solve_keplers_equation

  var lambda = OmegaBar_g + (2 * Math.atan(Math.sqrt((1+Eccentricity) / (1-Eccentricity)) * Math.tan(E/2)));
//End sun_ecliptic_longitude


//Compute ecliptic_to_equatorial(lambda, 0.0, alpha, delta)
  var sin_e = Math.sin(MeanObliquity);
  var cos_e = Math.cos(MeanObliquity);
  var alpha = Math.atan2(Math.sin(lambda)*cos_e, Math.cos(lambda));
  var delta = Math.asin(sin_e*Math.sin(lambda));
//End ecliptic_to_equatorial

//Compute GST(ssue)
  //Compute JD = julian_date(year, month, day)
    var TmpDate = new Date(ThisMilliseconds);
	var OffsetMilliseconds = ThisMilliseconds ; //+ (60000*TmpDate.getTimezoneOffset());
    var ThisDate = new Date(OffsetMilliseconds);
    var y = ThisDate.getYear();
      if (y < 1000) y += 1900;
    var m = ThisDate.getMonth() + 1;
    var z = ThisDate.getDate();
    
//	if ((m==1) || (m==2)) {
//	  y -= 1;
//	  m += 12; }

	var A = y / 100;
	var B = 2 - A + (A/4);
	var C = 365.25 * y;
	var D = 30.6001 * (m+1);

	var JD = B + C + D + z + 1720994.5;
  //End julian_date

  var T = (JD - 2451545) / 36525;
  var T0 = ((((T + 2.5862E-5) * T) + 2400.051336) * T) + 6.697374558;

  T0 = T0 % 24;
  if (T0 < 0) T0 += 24;

  var UT = ThisDate.getHours() + ((ThisDate.getMinutes() + (ThisDate.getSeconds() / 60)) / 60);
 
  T0 += UT * 1.002737909;
  T0 = T0 % 24;
  if (T0 < 0) T0 += 24;
//End GST

var tmp = alpha - ((TWOPI/24)*T0);
while (tmp < -Math.PI) {tmp += TWOPI;};
while (tmp > Math.PI) {tmp -= TWOPI;};

var lon = tmp * (360/TWOPI);
var lat = delta * (360/TWOPI);

//Generate the path of the appropriate xearth image

//lon is even
lon = Math.round(lon);
//if (lon & 1 != 0) {
if (lon % 2 != 0) {
  if (lon > 0) {
    lon -= 1;
    }
  else {
    lon += 1;
    }
  }
  
  // force even for mozilla 4.01-4.05
  // Consolidate lines, 4/5/01 - PRF
  lon = Math.round(lon/2) * 2;
  
if (lon <= -181) lon = -180;
if (lon >= 181) lon = 180;

//lat is odd
lat = Math.round(lat);
//if (lat & 1 == 0) {
if (lat % 2 == 0) {
  if (lat > 0) {
    lat -= 1;
    }
  else {
    lat += 1;
    }
  }
  
  // force odd for mozilla 4.01-4.05
  // Fixed broken algorithm, 4/5/01 - PRF
  lat =  (Math.round(lat/2 - 1) * 2) + 1;
  
  
if (lat <= -24) lat = -23;
if (lat >= 24) lat = 23;

var latStr; var lonStr;
if (lat < 0) 
  latStr = (-lat) + "S";
else 
  latStr = lat + "N";

if (lon < 0) 
  lonStr = (-lon) + "S";
else 
  lonStr = lon + "N";

return "images/xearths/" + latStr + "/" + lonStr + ".jpg";
//return lon;
}

File: python_20240425a/python/vb2pyconv/vb2py/PythonCard/samples/worldclock/xearth.py
import math, time

def getLatLong():
    "Returns URL for day/night picture"
    # Define some 'constants'
    ClientRecieveTime=time.time() * 1000
    # QueryTimeZone = 10
    QueryTimeZone = -time.timezone/3600
    QueryTimeZoneOffsetMin = QueryTimeZone * 60
    NISTSendTimeGMTms = ClientRecieveTime - ( QueryTimeZoneOffsetMin * 60 * 1000 )

    # NIST start time in the query time zone
    QueryTimeZoneOffset = ( QueryTimeZoneOffsetMin * 60 * 1000 )
    # Replace this with "QueryTimeZoneOffset = ( -time.timezone * 60 * 1000 )"

    # Client start time in some time zone
    ClientRecieveTimems = ClientRecieveTime

    # what timezone does your computer think it is? - in minutes
    ClientTimeZone = QueryTimeZoneOffsetMin;
    ClientNISTDelta = math.floor(NISTSendTimeGMTms - ClientRecieveTimems)

    currTime = ClientRecieveTime + ClientNISTDelta # ThisMilliseconds
    gmtTime = time.gmtime()
    ssue = currTime / 1000
    TwoPi = 2 * math.pi
    EpochStart = 631065600
    DaysSinceEpoch = (ssue - EpochStart)/ (24*3600)
    RadsPerDay = TwoPi / 365.242191
    Epsilon_g = 279.403303 * (TwoPi / 360)
    OmegaBar_g = 282.768422 * (TwoPi / 360)
    Eccentricity = 0.016713
    MeanObliquity = 23.440592 * (TwoPi / 360);
    # Calculate sun_ecliptic_longitude
    N = RadsPerDay * DaysSinceEpoch
    N = N % TwoPi
    if N < 0:
        N += TwoPi # This should never be executed, but never mind
    M_Sun = N + Epsilon_g - OmegaBar_g
    if M_Sun < 0:
        M_Sun += TwoPi # This should never be executed either
    # Now we solve keplers equation. For those who are interested keplers 
    # equation is all about plotting the orbit of an object on the 
    # elliptic plane.
    E = M_Sun
    while 1:
        delta = E - (Eccentricity*math.sin(E)) - M_Sun
        if (abs(delta) <= 1E-10):
            break
        E -= delta / (1 - (Eccentricity*math.cos(E)))
    # End of the keplers equation solution
    myLambda = OmegaBar_g + (2 * math.atan(math.sqrt((1+Eccentricity) / (1-Eccentricity)) * math.tan(E/2)))
    # There, finished calculating the sun ecliptic longitude
    # Now we calculate the ecliptic to equatorial (something or other)
    sin_e = math.sin(MeanObliquity)
    cos_e = math.cos(MeanObliquity)
    alpha = math.atan2(math.sin(myLambda)*cos_e, math.cos(myLambda))
    delta = math.asin(sin_e*math.sin(myLambda));
    # End of ecliptic to equatorial
    # We calculate the Julian date here, Python could probably do this better
    # I leave it to the casual observer to replace the following few lines
    y = gmtTime[0] # Year
    m = gmtTime[1] # Month number
    z = gmtTime[2] # Day number
    A = y / 100
    B = 2 - A + (A/4)
    C = 365.25 * y
    D = 30.6001 * (m+1)
    JD = B + C + D + z + 1720994.5
    T = (JD - 2451545) / 36525
    T0 = ((((T + 2.5862E-5) * T) + 2400.051336) * T) + 6.697374558
    T0 = T0 % 24
    if T0 < 0:
        T0 += 24
    UT = (float(gmtTime[3])) + ((float(gmtTime[4]) + (float(gmtTime[5]) / 60)) / 60)
    T0 += UT * 1.002737909
    T0 = T0 % 24
    if T0 < 0:
        T0 += 24

    tmp = alpha - ((TwoPi/24)*T0);
    while tmp < -math.pi:
        tmp += TwoPi
    while tmp > math.pi:
        tmp -= TwoPi

    # Now calculate our longitude and latitude
    lon = tmp * (360/TwoPi)
    lat = delta * (360/TwoPi)

    # Generate the path of the appropriate xearth image
    lon = round(lon)
    
    if (lon % 2 != 0):
      if (lon > 0):
        lon -= 1
      else:
        lon += 1

    lon = round(lon/2) * 2
    if lon <= -181:
        lon = -180
    if lon >= 181:
        lon = 180
    # lat is odd
    lat = round(lat)
    if (lat % 2 == 0):
        if lat > 0:
            lat -= 1
        else:
            lat += 1
    # Need to do different calculations for negative and positive values of lat
    # to emulate the way javascript handles rounding
    if lat < 0:
        lat = (round(int(lat/2) - 1) * 2) + 1
    else:
        lat = (round(lat/2 - 1) * 2) + 1
  
    if lat <= -24:
        lat = -23
    if lat >= 24:
        lat = 23

    if lat < 0:
        latStr = str(int(-lat)) + "S"
    else:
        latStr = str(int(lat)) + "N"
    if lon < 0:
        lonStr = str(int(-lon)) + "S"
    else:
        lonStr = str(int(lon)) + "N"
    # url = "http://www.time.gov/" + getLatLong()
    # return 'http://www.time.gov/images/xearths/11N/154N.jpg'
    return "images/xearths/" + latStr + "/" + lonStr + ".jpg"
